network/p2p/p2pnet/network.go

```diff
@@ -0,0 +1,1266 @@
+func WithAlspConfig(cfg *alspmgr.MisbehaviorReportManagerConfig) NetworkConfigOption {
+	return func(params *NetworkConfig) {
+		params.AlspCfg = cfg
+	}
+}
+
+// WithCodec overrides the default codec (i.e., encoder and decoder). It is mostly used for testing purposes.
+// Note: do not override the default codec in production unless you know what you are doing.
+func WithCodec(codec network.Codec) NetworkConfigOption {
+	return func(params *NetworkConfig) {
+		params.Codec = codec
+	}
+}
+
+func WithSlashingViolationConsumerFactory(factory func(adapter network.Adapter) network.ViolationsConsumer) NetworkConfigOption {
+	return func(params *NetworkConfig) {
+		params.SlashingViolationConsumerFactory = factory
+	}
+}
+
+// NetworkOption is a function that can be used to override network attributes.
+// It is mostly used for testing purposes.
+// Note: do not override network attributes in production unless you know what you are doing.
+type NetworkOption func(*Network)
+
+// WithAlspManager sets the misbehavior report manager for the network. It overrides the default
+// misbehavior report manager that is created from the config.
+// Note that this option is mostly used for testing purposes, do not use it in production unless you
+// know what you are doing.
+//
+// Args:
+//
+//	mgr: misbehavior report manager
+//
+// Returns:
+//
+//	NetworkOption: network option
+func WithAlspManager(mgr network.MisbehaviorReportManager) NetworkOption {
+	return func(n *Network) {
+		n.misbehaviorReportManager = mgr
+	}
+}
+
+// WithPeerManagerFilters sets the peer manager filters for the network. It overrides the default
+// peer manager filters that are created from the config.
+func WithPeerManagerFilters(filters ...p2p.PeerFilter) NetworkOption {
+	return func(n *Network) {
+		n.peerManagerFilters = filters
+	}
+}
+
+// WithUnicastRateLimiters sets the unicast rate limiters for the network. It overrides the default
+// unicast rate limiters that are created from the config.
+func WithUnicastRateLimiters(limiters *ratelimit.RateLimiters) NetworkOption {
+	return func(n *Network) {
+		n.unicastRateLimiters = limiters
+	}
+}
+
+// WithPreferredUnicastProtocols sets the preferred unicast protocols for the network. It overrides the default
+// preferred unicast.
+func WithPreferredUnicastProtocols(protocols ...protocols.ProtocolName) NetworkOption {
+	return func(n *Network) {
+		n.preferredUnicasts = protocols
+	}
+}
+
+// WithMessageValidators sets the message validators for the network. It overrides the default
+// message validators.
+func WithMessageValidators(validators ...network.MessageValidator) NetworkOption {
+	return func(n *Network) {
+		n.validators = validators
+	}
+}
+
+// NewNetwork creates a new network with the given configuration.
+// Args:
+// param: network configuration
+// opts: network options
+// Returns:
+// Network: a new network
+func NewNetwork(param *NetworkConfig, opts ...NetworkOption) (*Network, error) {
```

Comment:
The network serves as a comprehensive layer that integrates three interfaces within Flow. Renaming it as `NewNetworkEngine` would misrepresent its function, implying it's merely an engine component rather than an all-encompassing networking layer.